### PR TITLE
[Performance] Optimize DeepSeek-V4 VL embed_input_ids to remove per-step copies

### PR DESCRIPTION
**Pull request by frankie-ys**
### What this PR does / why we need it?

Profiling the DeepSeek-V4-Flash-Vision-Exp deployment (`dsv4-vision-exp`) showed ~4ms of copy operations inside `embed_input_ids` on every scheduler step. 
131a6c9c094b851c239743a38e356ca5

The cause was that the sentinel-replacement path re-executed avoidable copies per forward:

1. `torch.stack([...]).to(inputs_embeds.dtype)` — the four `image_*` parameters are float32 while embeddings are bf16, so every step re-stacked and re-cast the sentinel vectors.
2. `sentinel_mask & ~is_multimodal.to(input_ids.device)` — `is_multimodal` lives on CPU in this stack, so every step performed an H2D copy of the mask plus a device AND. This exclusion is unnecessary: positions flagged `is_multimodal` are always overwritten afterwards by `_merge_multimodal_embeddings`, so the table value written there is dead anyway.
3. `torch.where(...)` allocated a fresh full `seq_len x hidden` tensor per step on top of the `table[idx]` gather.

Changes:

- Cache the stacked sentinel table in the embedding dtype (`_get_sentinel_table`), built once per model instance and reused for the process lifetime.
- Drop the redundant `& ~is_multimodal` exclusion and its per-step H2D copy (safe because `_merge_multimodal_embeddings` overwrites those positions afterwards).
- Write the sentinel replacement in place via `torch.where(..., out=inputs_embeds)`, removing one full-sequence allocation per step.

After the change, `embed_input_ids` takes ~785us (down from ~4ms+), with the remaining cost being the inherent `table[idx]` gather + `where` memory traffic.
e3e4651a952f2c6339f95cc51949e53b

benchmark result:

origin:
    
    ============ Serving Benchmark Result ============
    Successful requests:                     100
    Failed requests:                         0
    Maximum request concurrency:             6
    Benchmark duration (s):                  385.63
    Total input tokens:                      5000400
    Total generated tokens:                  35000
    Request throughput (req/s):              0.26
    Output token throughput (tok/s):         90.76
    Peak output token throughput (tok/s):    78.00
    Peak concurrent requests:                9.00
    Total token throughput (tok/s):          13057.75
    ---------------Time to First Token----------------
    Mean TTFT (ms):                          2648.71
    Median TTFT (ms):                        1703.86
    P99 TTFT (ms):                           16038.56
    -----Time per Output Token (excl. 1st token)------
    Mean TPOT (ms):                          57.60
    Median TPOT (ms):                        56.45
    P99 TPOT (ms):                           108.71
    ---------------Inter-token Latency----------------
    Mean ITL (ms):                           127.53
    Median ITL (ms):                         90.50
    P99 ITL (ms):                            1168.05
    ----------------End-to-end Latency----------------
    Mean E2EL (ms):                          22751.24
    Median E2EL (ms):                        22026.70
    P99 E2EL (ms):                           41368.55
    ---------------Speculative Decoding---------------
    Acceptance rate (%):                     17.00
    Acceptance length:                       2.02
    Drafts:                                  17349
    Draft tokens:                            104094
    Accepted tokens:                         17700
    Per-position acceptance (%):
      Position 0:                            41.41
      Position 1:                            22.94
      Position 2:                            15.05
      Position 3:                            10.39
      Position 4:                            7.68      
      Position 5:                            4.56
    ==================================================
    
After optimization：

    ============ Serving Benchmark Result ============
    Successful requests:                     100
    Failed requests:                         0
    Maximum request concurrency:             6
    Benchmark duration (s):                  373.37
    Total input tokens:                      5000400
    Total generated tokens:                  35000
    Request throughput (req/s):              0.27
    Output token throughput (tok/s):         93.74
    Peak output token throughput (tok/s):    72.00
    Peak concurrent requests:                9.00
    Total token throughput (tok/s):          13486.31
    ---------------Time to First Token----------------
    Mean TTFT (ms):                          2589.93
    Median TTFT (ms):                        1673.65
    P99 TTFT (ms):                           16051.38
    -----Time per Output Token (excl. 1st token)------
    Mean TPOT (ms):                          56.25
    Median TPOT (ms):                        56.77
    P99 TPOT (ms):                           104.24
    ---------------Inter-token Latency----------------
    Mean ITL (ms):                           128.55
    Median ITL (ms):                         89.94
    P99 ITL (ms):                            1164.70
    ----------------End-to-end Latency----------------
    Mean E2EL (ms):                          22222.08
    Median E2EL (ms):                        22170.34
    P99 E2EL (ms):                           38360.14
    ---------------Speculative Decoding---------------
    Acceptance rate (%):                     18.00
    Acceptance length:                       2.08
    Drafts:                                  16848
    Draft tokens:                            101088
    Accepted tokens:                         18199
    Per-position acceptance (%):
      Position 0:                            40.70
      Position 1:                            24.25
      Position 2:                            16.93
      Position 3:                            11.47
      Position 4:                            8.88
      Position 5:                            5.79
    ==================================================

### Does this PR introduce _any_ user-facing change?

No. The model's outputs are unchanged: the cached table depends only on weights fixed after loading; the in-place `where` writes bit-identical values; positions excluded by the removed mask are overwritten by the multimodal merge regardless. This is purely a performance refactor of a hot path.

### How was this patch tested?

- Added unit tests in `tests/ut/models/test_deepseek_v4_vision.py`:
  - sentinel positions (start/pad/image/newline/end) are replaced from the cached table, regular tokens pass through unchanged;
  - the table is cached (identity preserved across calls) in bf16;
  - passthrough when the vision component is disabled.

- vLLM main: https://github.com/vllm-project/vllm/commit/84030bbe3d74d99bad477a3d2e37a973ccd8865c

### CODE REVIEW

**Comment by kunpengW-code** [comment]
Sorry, this PR may not be able to move forward. There are no ongoing plans for further optimization of the DeepSeek-V4-Flash-Vision-Exp model. Our main focus will shift to V4.1-flash.

**Comment by frankie-ys** [comment]
> Sorry, this PR may not be able to move forward. There are no ongoing plans for further optimization of the DeepSeek-V4-Flash-Vision-Exp model. Our main focus will shift to V4.1-flash.

Thank you for your reply. However, since we are currently unable to deploy dsv4.1 on a single 910B machine—and this optimization would boost the throughput of dsv4-vision-exp—I will follow up on the dsv4.1 implementation once the w4 dsv4.1 weights become available.
